ocaml/sdk-gen/go/gen_go_helper.ml

```diff
@@ -0,0 +1,274 @@
+   it under the terms of the GNU Lesser General Public License as published
+   by the Free Software Foundation; version 2.1 only. with the special
+   exception on linking described in file LICENSE.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+*)
+
+(* Generator of Go bindings from the datamodel *)
+
+open Datamodel_types
+open CommonFunctions
+
+let templates_dir = "templates"
+
+let ( // ) = Filename.concat
+
+let snake_to_camel (s : string) : string =
+  String.split_on_char '_' s
+  |> List.map (fun s -> String.split_on_char '-' s)
+  |> List.concat
+  |> List.map String.capitalize_ascii
+  |> String.concat ""
+
+let render_template template_file json =
+  let templ =
+    string_of_file (templates_dir // template_file) |> Mustache.of_string
+  in
+  Mustache.render templ json
+
+let generate_file rendered destdir output_file =
+  let out_chan = open_out (destdir // output_file) in
+  Fun.protect
+    (fun () -> output_string out_chan rendered)
+    ~finally:(fun () -> close_out out_chan)
+
+module EnumSet = Set.Make (struct
+  type t = string * (string * string) list
+
+  let compare (x0, _y0) (x1, _y1) = String.compare x0 x1
+end)
+
+module Json = struct
+  type enum = (string * string) list
+
+  module StringMap = Map.Make (String)
+
+  type enums = enum StringMap.t
+
+  let choose_enum _key a _b = Some a
+
+  let rec string_of_ty_with_enums ty : string * enums =
+    match ty with
+    | SecretString | String ->
+        ("string", StringMap.empty)
+    | Int ->
+        ("int", StringMap.empty)
+    | Float ->
+        ("float64", StringMap.empty)
+    | Bool ->
+        ("bool", StringMap.empty)
+    | DateTime ->
+        ("time.Time", StringMap.empty)
+    | Enum (name, kv) ->
+        let name = snake_to_camel name in
+        (name, StringMap.singleton name kv)
+    | Set ty ->
+        let s, e = string_of_ty_with_enums ty in
+        ("[]" ^ s, e)
+    | Map (ty1, ty2) ->
+        let s1, e1 = string_of_ty_with_enums ty1 in
+        let s2, e2 = string_of_ty_with_enums ty2 in
+        let ty = "map[" ^ s1 ^ "]" ^ s2 in
+        (ty, StringMap.union choose_enum e1 e2)
+    | Ref r ->
+        (snake_to_camel r ^ "Ref", StringMap.empty)
+    | Record r ->
+        (snake_to_camel r ^ "Record", StringMap.empty)
+    | Option ty ->
+        string_of_ty_with_enums ty
+
+  let of_enum name vs =
+    let name = snake_to_camel name in
+    let of_value (v, d) =
+      `O
+        [
+          ("value", `String v)
+        ; ("doc", `String d)
+        ; ("name", `String (name ^ snake_to_camel v))
+        ; ("type", `String name)
+        ]
+    in
+    `O [("name", `String name); ("values", `A (List.map of_value vs))]
+
+  let fields_of_obj_with_enums obj =
```

Comment:
How about to split this function into the following:
`fields_of_obj`
`of_field`
`enums_of_field`

```

let fields = fields_of_obj obj in
let enums =
  List.map enums_of_field fields
  |> List.concat
  |> List.fold_left
    (fun acc enum ->
      let enums = enums_of_field field in
      StringMap.union pick_first acc enum
    )
    StringMap.empty fields
in
let types = List.map (fun field -> field.ty) fields in
let enums_json = List.map of_enum enums
let fields_json = List.map of_field fields in
...
```